Project/DAQ_System/Program/Src/app.cpp

```diff
@@ -133,30 +111,32 @@ void cppMain() {
 
 
 
-		//HAL_Delay(150);
-		if(can_bus->MessageArrivedFlag()){
+		if (ecu->NewMessageArrived()) {
 			__disable_irq();
-//			can_bus->DisableInterruptMode();
-			can_bus->Receive(rx_buffer);
 
-			if(can_bus->LatestCanId() == 0x0CFFF148) {
-				sensor::FramePE2 frame(rx_buffer);
-				float manifold_absolute_pressure = frame.ManifoldAbsolutePressure();
+			ecu->Update();
+			uint32_t can_id = ecu->LatestCanId();
+
+			switch(can_id) {
+			case FramePe2Id:
+				manifold_absolute_pressure = ecu->map_;
 
-				printf("\t\t 0x0CFFF148 \n");
+				printf("\t\t %" PRIu32 "\n", can_id);
 				printf("Manifold Pressure: %f\n", manifold_absolute_pressure);
 				printf("\r");
+				break;
 
-			} else if(can_bus->LatestCanId() == 0x0CFFF548){
-				printf("\t\t 0x0CFFF548 \n");
-				printf("\r");
-				// Process raw data
+
+			case FramePe6Id:
+				printf("\t\t %" PRIu32 "\n", can_id);
+				break;
 			}
```

Comment:
As I develop integrate the `IEcu` interface, I am starting to realize that in practice we may interact with the `Pe3` ECU class directly (as shown in this snippet).

Interacting through `IEcu`, we will lose the ability to identify which data from ECU got updated:
```c++

if (generic_ecu->NewMessageArrived()) {
  generic_ecu->Update();
  // get data
}
```

Notice that in this example snippet, I am not using the switch statement to handle specific CAN IDs. The IDs are ECU-specific. So, although I can use `generic_ecu->LatestCanId()` to retrieve the ID, identifying it will require ECU-specific knowledge that is held in `Pe3`.

The example snippet still works, but identifying the CAN message will be more efficient. The example is essentially polling `Pe3` through `IEcu`. On the other hand, with the switch statement of ECU-specific CAN IDs, we would only be retrieving the data that was updated — instead of retrieving all of it.

**Long story short: `IEcu` would only support polling. While using `Pe3` directly will allow us to reap the benefits of an interrupt approach.**

So, in practice, I think we will have the DAQ use `Pe3` directly, and maintain `IEcu` for aiding developers to implement the minimum requirements of an ECU. But, the DAQ firmware will not interact to the ECU component using `IEcu`, only its concrete classes.

Reply:
A solution may be to implement the [Observer design pattern](https://www.modernescpp.com/index.php/the-observer-pattern/). But I think that outside the scope of this task and there are higher priorities.

Reply:
Updated `Pe3` usage, using its ecu-specific CAN IDs to act upon only the data that was recently updated:

![image](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/71054319/0bdc8fdc-0821-46be-83c7-24bcc84fb391)

Updated `IEcu` usage, where CAN ID identification is lost — so everything must be polled:

![image](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/71054319/3694e461-4f47-495a-8855-2d2a82dc5ff3)